feat: Add ```/rerank``` API for NVIDIA Inference Provider

jiayin-nvidia opened the pull request:
# What does this PR do?
Add rerank API for NVIDIA Inference Provider.

Related PR for supporting rerank on the client side: [https://github.com/llamastack/llama-stack-client-python/pull/271](https://github.com/llamastack/llama-stack-client-python/pull/271)

<!-- If resolving an issue, uncomment and update the line below -->
<!-- Closes #3278  -->

## Test Plan
TODO: Add tests and example documentation.

jiayin-nvidia commented:
Hi @ashwinb @franciscojavierarceo @mattf @ehhuang, the rerank integration tests for this PR needs the client-side rerank support ([https://github.com/llamastack/llama-stack-client-python/pull/271](https://github.com/llamastack/llama-stack-client-python/pull/271)) to run successful. Could you please also review the client-side changes? I can also delete the rerank integration tests for now and add them in a separate PR after the client-side PR is merged and released. Please let me know which approach you prefer. Thank you!

franciscojavierarceo commented on llama_stack/providers/remote/inference/nvidia/nvidia.py:
```diff
@@ -209,6 +215,80 @@ async def embeddings(
+            "Content-Type": "application/json",
+        }
+
+        import aiohttp
```
why keep the import here?

jiayin-nvidia commented on docs/source/providers/agents/index.md:
```diff
@@ -4,12 +4,12 @@
-    - Agents can be provided with various tools (see the ToolGroups and ToolRuntime APIs for more details).
-    - Agents can be provided with various shields (see the Safety API for more details).
-    - Agents can also use Memory to retrieve information from knowledge bases. See the RAG Tool and Vector IO APIs for more details.
+Main functionalities provided by this API:
```
Those were automatically changed by the pre-commit hooks. I can remove the changes.

jiayin-nvidia commented on tests/unit/providers/nvidia/test_rerank_inference.py:
```diff
@@ -0,0 +1,180 @@
+        return PostContext(self.response)
+
+
+def create_adapter(config=None, model_metadata=None):
```
This is doing the unit test. The integration test is under ```tests/integration/inference/test_rerank.py```

jiayin-nvidia replied:
No problem, thanks for your review!

jiayin-nvidia commented on docs/source/providers/batches/index.md:
```diff
@@ -3,15 +3,15 @@
 The Batches API enables efficient processing of multiple requests in a single operation,
-    particularly useful for processing large datasets, batch evaluation workflows, and
-    cost-effective inference at scale.
+particularly useful for processing large datasets, batch evaluation workflows, and
```
Will remove the changes

franciscojavierarceo commented on docs/source/providers/inference/index.md:
```diff
@@ -7,6 +7,7 @@ Llama Stack Inference API for generating completions, chat completions, and embe
     This API provides the raw interface to the underlying models. Two kinds of models are supported:
     - LLM models: these models generate "raw" and "chat" (conversational) completions.
     - Embedding models: these models generate embeddings to be used for semantic search.
+    - Rerank models: these models rerank the documents by relevance. 
```
looks like the precommit didn't like this. can you run `uv run pre-commit run --all-files`?

ehhuang commented on tests/integration/inference/test_rerank.py:
```diff
@@ -0,0 +1,147 @@
+from llama_stack.core.library_client import LlamaStackAsLibraryClient
+
+# Test data
+DUMMY_STRING = "string_1"
+DUMMY_STRING2 = "string_2"
+DUMMY_TEXT = TextContentItem(text=DUMMY_STRING, type="text")
+DUMMY_TEXT2 = TextContentItem(text=DUMMY_STRING2, type="text")
+DUMMY_IMAGE_URL = ImageContentItem(
+    image=ImageContentItemImage(url=ImageContentItemImageURL(uri="https://example.com/image.jpg")), type="image"
+)
+DUMMY_IMAGE_BASE64 = ImageContentItem(image=ImageContentItemImage(data="base64string"), type="image")
```
can we rank some obvious (query, text/images) pairs to validate correctness?

jiayin-nvidia replied:
I added several text query and text items pairs example. The image query and image items are not supported right now.

franciscojavierarceo commented:
@jiayin-nvidia pre-commit failed, can you fix and update?

ehhuang commented on llama_stack/providers/remote/inference/nvidia/nvidia.py:
```diff
@@ -210,6 +217,78 @@ async def embeddings(
+        # Convert query to text format
+        if isinstance(query, str):
+            query_text = query
+        elif hasattr(query, "text"):
```
just check for OpenAIChatCompletionContentPartTextParam explicitly?

ehhuang commented on llama_stack/providers/remote/inference/nvidia/nvidia.py:
```diff
@@ -210,6 +217,78 @@ async def embeddings(
+        for item in items:
+            if isinstance(item, str):
+                passages.append({"text": item})
+            elif hasattr(item, "text"):
```
ditto

ehhuang reviewed (commented):
LG overall. one small comment and CI failures

jiayin-nvidia commented:
Thank you @ehhuang , comments are addressed and the CI is passing. I also updated the OpenAI Mixin to support ```rerank_model_list``` and add rerank models as static models in NVIDIA inference provider due to recent removal of the model list for the provider. I also updated the rerank integration tests based on recent client side changes.

franciscojavierarceo reviewed (approved):
overall this lgtm, thanks so much @jiayin-nvidia for this!

jiayin-nvidia commented:
> @jiayin-nvidia can you resolve the conflicts please?

@franciscojavierarceo All conflicts are resolved and CIs are passing. Please let me know if anything else is needed or if it's ready to be merged. Thank you!

jiayin-nvidia commented:
Hi @franciscojavierarceo @ehhuang , I would like to follow up on this PR to see if anything is needed before it is merged. Thank you!
